src/pages/iou/steps/MoneyRequestAmountForm.js

```diff
@@ -0,0 +1,266 @@
+    amount: 0,
+    currency: CONST.CURRENCY.USD,
+    forwardedRef: null,
+    isEditing: false,
+};
+
+/**
+ * Returns the new selection object based on the updated amount's length
+ *
+ * @param {Object} oldSelection
+ * @param {Number} prevLength
+ * @param {Number} newLength
+ * @returns {Object}
+ */
+const getNewSelection = (oldSelection, prevLength, newLength) => {
+    const cursorPosition = oldSelection.end + (newLength - prevLength);
+    return {start: cursorPosition, end: cursorPosition};
+};
+
+const AMOUNT_VIEW_ID = 'amountView';
+const NUM_PAD_CONTAINER_VIEW_ID = 'numPadContainerView';
+const NUM_PAD_VIEW_ID = 'numPadView';
+
+function MoneyRequestAmountForm({amount, currency, isEditing, forwardedRef, onCurrencyButtonPress, onSubmitButtonPress}) {
+    const {translate, toLocaleDigit, fromLocaleDigit, numberFormat} = useLocalize();
+
+    const textInput = useRef(null);
+
+    const selectedAmountAsString = amount ? CurrencyUtils.convertToWholeUnit(currency, amount).toString() : '';
+
+    const [currentAmount, setCurrentAmount] = useState(selectedAmountAsString);
+    const [shouldUpdateSelection, setShouldUpdateSelection] = useState(true);
+
+    const [selection, setSelection] = useState({
+        start: selectedAmountAsString.length,
+        end: selectedAmountAsString.length,
+    });
+
+    /**
+     * Event occurs when a user presses a mouse button over an DOM element.
+     *
+     * @param {Event} event
+     * @param {Array<string>} nativeIds
+     */
+    const onMouseDown = (event, nativeIds) => {
+        const relatedTargetId = lodashGet(event, 'nativeEvent.target.id');
+        if (!_.contains(nativeIds, relatedTargetId)) {
+            return;
+        }
+        event.preventDefault();
+        if (!textInput.current) {
+            return;
+        }
+        if (!textInput.current.isFocused()) {
+            textInput.current.focus();
+        }
+    };
+
+    /**
+     * Convert amount to whole unit and update selection
+     *
+     * @param {String} currencyCode
+     * @param {Number} amountInCurrencyUnits
+     */
+    const saveAmountToState = (currencyCode, amountInCurrencyUnits) => {
+        if (!currencyCode || !amountInCurrencyUnits) {
+            return;
+        }
+        const amountAsStringForState = CurrencyUtils.convertToWholeUnit(currencyCode, amountInCurrencyUnits).toString();
+        setCurrentAmount(amountAsStringForState);
+        setSelection({
+            start: amountAsStringForState.length,
+            end: amountAsStringForState.length,
+        });
+    };
+
+    useEffect(() => {
+        saveAmountToState(currency, amount);
+        // we want to update the state only when the amount is changed
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [amount]);
+
+    /**
+     * Sets the state according to amount that is passed
+     * @param {String} newAmount - Changed amount from user input
+     */
+    const setNewAmount = (newAmount) => {
```

Comment:
NAB, I wonder if we can make all the methods related to amounts more descriptive. What's the difference between "setting the new amount" vs. "setting the current amount" vs. "updating the amount" vs. "saving the amount to state". I still feel like we have some really heavy logic all related to the "amount" here and that was one of the main things I would still like to see addressed as part of this code cleanup.

Reply:
I'd move this addLeadingZero to TextInputWithCurrencySymbol (and updateAmount could be removed) , and remove saveAmountToState -> it’s only used in one useEffect, it can be done inside the hook